### PR TITLE
Use fetchAndPersistBracket instead of fetchAndPersist

### LINKED ISSUE #11158: use fetchAndPersistBracket instead of fetchAndPersist
#10617 mentions that `fetchAndPersistBracket` can be used for the non-websocket use cases where plain `fetchAndPersist` is currently used. It's probably a broadly good idea. The task here is to port every caller of the latter to the former, following [the current example](https://github.com/digital-asset/daml/commit/b4d00317b402bee59a06f77a41ad8fad8a173e66#diff-7e56e95d3c90eb765c9ce924c97401736293cd19ceb284872b2a74e18acdf57fL716-L727) as well as the advice of #11157, and then make `fetchAndPersist` a `private[this]` function.

### PR DESCRIPTION
**Pull request by skisel**
Fixes #11158

```rst
CHANGELOG_BEGIN
- [JSON API] Under rare conditions, a POST-style multi-template query backed by
  database could have different template IDs appear to be queried at different
  ledger times, if updated concurrently.  This conditions is now checked and
  accounted for.  See `issue #12672
  <https://github.com/digital-asset/daml/pull/12672>`__.
CHANGELOG_END
```

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala [hunk 1]
```diff
@@ -325,15 +325,15 @@ class ContractsService(
...
               timed(
                 metrics.daml.HttpJsonApi.Db.fetchByIdFetch,
-                fetch.fetchAndPersist(jwt, ledgerId, parties, List(resolved)),
-              ) *>
-                timed(
-                  metrics.daml.HttpJsonApi.Db.fetchByIdQuery,
-                  ContractDao.fetchById(parties, resolved, contractId),
-                )
+                fetch.fetchAndPersistBracket(jwt, ledgerId, parties, List(resolved)) { _ =>
+                  timed(
```
Comment (skisel) [inline ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala:331]: Not sure about nested `timed` here, how does it work? is it fine?

Comment (S11001001) [reply ledger-service/http-json/src/main/scala/com/digitalasset/http/ContractsService.scala:331]: The trouble is that what we are wanting to measure here is the difference between time fetching for the query, and time actually doing the query (i.e. the fetchById call), and conflating the two would be _really bad_ for the accuracy of our metrics. The best solution here might be to make `fetchAndPersistBracket` a little more complicated, possibly by having it take a polymorphic function `ConnectionIO ~> ConnectionIO` as an argument (it seems worse to have it talk about metrics itself).

**Comment by S11001001** [comment]
@realvictorprm `fetchAndPersistBracket` #10617 changes our approach from "do all fetching, then query" to "fetch, then query, then check consistency, then if needed catch-up fetch, then query..."

What is the most useful way to present these metrics? There's some case to be made for aggregating fetch time for a particular request, since _either_ we roll back _or_ keep the prior fetch and catch-up depending on how recoverable the inconsistency was. It's more difficult to justify aggregating query time, since we always discard all query results if the consistency check fails and only ever use the last result. On the other hand, the time for those discarded queries _does_ contribute to the request time; there is no way to parallelize it with the "good" query.

Should we record each fetch separately? Add them all up? Discard or keep the bad queries? Spend any time at all tabulating the consistency check (it is "just" a select on the `ledger_offsets` table)?

**Review by S11001001 — APPROVED** [review]

**Comment by S11001001** [comment]
@skisel I reverted the contract ID and contract key lookups; explanation in comments.

**Pull request merged**